source/core_mqtt_serializer.c

```diff
@@ -1286,17 +1289,17 @@ static MQTTStatus_t deserializePublish( const MQTTPacketInfo_t * pIncomingPacket
     {
         /* Calculate the length of the payload. QoS 1 or 2 PUBLISH packets contain
          * a packet identifier, but QoS 0 PUBLISH packets do not. */
-        if( pPublishInfo->qos == MQTTQoS0 )
-        {
-            pPublishInfo->payloadLength = ( pIncomingPacket->remainingLength - pPublishInfo->topicNameLength - sizeof( uint16_t ) );
-            pPublishInfo->pPayload = pPacketIdentifierHigh;
-        }
-        else
+        pPublishInfo->payloadLength = pIncomingPacket->remainingLength - pPublishInfo->topicNameLength - sizeof( uint16_t );
+
+        if( pPublishInfo->qos != MQTTQoS0 )
         {
-            pPublishInfo->payloadLength = ( pIncomingPacket->remainingLength - pPublishInfo->topicNameLength - 2U * sizeof( uint16_t ) );
-            pPublishInfo->pPayload = pPacketIdentifierHigh + sizeof( uint16_t );
+            /* Two more bytes for the packet identifier. */
+            pPublishInfo->payloadLength -= sizeof( uint16_t );
```

Comment:
Instead of adding two bytes in advance then subtracting if `pPublishInfo->qos != MQTTQoS0`, why not just add once if `pPublishInfo->qos != MQTTQoS0`?

Reply:
Not sure what you mean, we're not adding any bytes to the payload length. The payload length is defined as: Remaining length - # of Topic length bytes - Topic length - # of packet ID bytes. There are two packet ID bytes for QoS > 0, and 0 none for QoS 0. So, we need to subtract an additional 2 for QoS > 0.

Reply:
Oh got it, I seem to have misread the code.